### PR TITLE
Support STRTOK, STRTOK_TO_ARRAY, SPLIT in BodoSQL C++ backend

### PR DESCRIPTION
**Pull request by dmsa-or**
## Changes included in this PR

Support STRTOK, STRTOK_TO_ARRAY, and SPLIT in C++ backend.
Add `get_var_py_args_as_types` utility function. This is like `get_py_args_as_types` but instead of accepting a fixed number of arguments, it can accept a range of possible argument counts. This is useful if the Arrow compute function has multiple options, some of which have default values. Instead of needing a separate call to `get_py_args_as_types` for each possible number of optional arguments that might be provided, you can call `get_var_py_args_as_types` once and process the resulting values without worrying anymore about argument count.

## Testing strategy

PR CI and Nightly.

## Checklist
- [X] PR title contains "[GPU]" if changes target Bodo DataFrames GPU acceleration.
- [x] Pipelines passed before requesting review. To run CI you must include `[run CI]` in your commit message.
- [X] I am familiar with the [Contributing Guide](https://github.com/bodo-ai/Bodo/blob/main/CONTRIBUTING.md)
- [X] I have installed + ran pre-commit hooks.

### CODE REVIEW

File: bodo/pandas/_util.h [hunk 1]
```diff
@@ -570,16 +572,152 @@ auto get_py_args_as_types(PyObject *args, const char *func_name,
...
+ * of converters should be equal to the maximum allowed argument count.
+ * @return A tuple where elements [0, MinRequired) are the first MinRequired
+ * Python arguments that have been converted by the respective `converters`, and
+ * elements [MinRequired, len(Converters)) are each an std::optional
+ * representing the converted Python argument if it exists, or std::nullopt if
+ * not.
+ */
+template <size_t MinRequired, typename... Converters>
+auto get_var_py_args_as_types(PyObject *args, const char *func_name,
+                              const std::vector<size_t> &allowed_arg_counts,
```
Comment (IsaacWarren) [inline bodo/pandas/_util.h]: I think this could be a constexpr std::array so we can validate arg counts at compile time

File: bodo/pandas/_util.h [hunk 2]
```diff
@@ -570,16 +572,152 @@ auto get_py_args_as_types(PyObject *args, const char *func_name,
...
+        }
+        return std::optional<result_t>(converter(arg, func_name));
+    }
+}
+
+/* Implementation detail of `get_var_py_args_as_types` */
+template <size_t MinRequired, typename... Converters, std::size_t... Is>
+auto _get_var_py_args_as_types_tuple(PyObject *args, const char *func_name,
+                                     std::index_sequence<Is...>,
+                                     Converters... converters) {
```
Comment (IsaacWarren) [inline bodo/pandas/_util.h:621]: Can you leave a comment explaining this, I had to lookup the syntax

File: bodo/pandas/_util.h [hunk 3]
```diff
@@ -570,16 +572,152 @@ auto get_py_args_as_types(PyObject *args, const char *func_name,
...
+    size_t max_allowed =
+        *std::max_element(allowed_arg_counts.begin(), allowed_arg_counts.end());
+    if (max_allowed > sizeof...(Converters)) {
+        throw std::runtime_error(
+            fmt::format("get_var_py_args_as_types: The passed maximum allowed "
+                        "arg count of {} for {} is greater than the number of "
+                        "provided PyObject-to-type conversion functions ({})",
+                        max_allowed, func_name, sizeof...(Converters)));
+    } else if (max_allowed < sizeof...(Converters)) {
+        std::cerr
```
Comment (IsaacWarren) [inline bodo/pandas/_util.h]: This should be a runtime error?

File: bodo/pandas/_util.h [hunk 4]
```diff
@@ -570,16 +572,152 @@ auto get_py_args_as_types(PyObject *args, const char *func_name,
...
+    if (std::find(allowed_arg_counts.begin(), allowed_arg_counts.end(),
+                  arg_count) == allowed_arg_counts.end()) {
+        // Format the list of allowed arg counts nicely for the error message
+        std::string allowed_counts_str;
+        for (size_t count : allowed_arg_counts) {
+            if (!allowed_counts_str.empty()) {
+                allowed_counts_str += ", ";
+            }
+            allowed_counts_str += std::to_string(count);
+        }
```
Comment (IsaacWarren) [inline bodo/pandas/_util.h:695]: I think you can do this with fmt, no need to change it though

File: BodoSQL/bodosql/tests/test_string_ops_first_half.py [hunk 5]
```diff
@@ -1083,6 +1090,12 @@ def test_split(query, memory_leak_check):
             }
         )
     }
+    # NOTE: We are comparing against DuckDB here, but DuckDB has slightly different
+    # semantics than Snowflake in some edge cases that are not represented in
+    # this test. For example, when the separator is empty, DuckDB splits the string
+    # into a list of its characters, whereas Snowflake just wraps the string in a
+    # list. DuckDB does wrap the original string in a list when the separator is
+    # NULL, but Snowflake returns NULL in that case.
```
Comment (IsaacWarren) [inline BodoSQL/bodosql/tests/test_string_ops_first_half.py]: Can we add tests with these cases with expected_output instead of check_query?

**Review by IsaacWarren — APPROVED** [review]

File: bodo/pandas/_util.h (hunk 3 above)
Comment (dmsa-or) [reply bodo/pandas/_util.h]: It was meant to be more of a warning but I can make it a runtime error.

File: bodo/pandas/_util.h [hunk 6]
```diff
@@ -570,16 +572,146 @@ auto get_py_args_as_types(PyObject *args, const char *func_name,
...
+            get_py_arg(Is), func_name, converters)...);
+}
+
+/**
+ * @brief Get the PyTuple of function arguments as a tuple of the arguments
+ * cast to the requested types specified by a sequence of type conversion
+ * functions. Later arguments beyond the MinRequired template parameter
+ * are wrapped in std::optional.
+ *
+ * @tparam MinRequired The minimum number of arguments that are always required
```
Comment (ehsantn) [inline bodo/pandas/_util.h]: Isn't `MinRequired` redundant?

**Review by ehsantn — APPROVED** [review]
Thanks @dmsa-or.

File: bodo/pandas/_util.h (hunk 6 above)
Comment (dmsa-or) [reply bodo/pandas/_util.h]: Oh yeah, I think it is now. Before I had `AllowedArgCounts` as a regular vector so I still needed MinRequired. But now I should be able to get rid of that.
